### DISCUSSION TITLE
how can i use it to give the list in dmenu?

### DISCUSSION THREAD

**Discussion by semi710** [Slackadays/Clipboard#116] [Q&A]
is there any way i can list the clipboard content in a dmenu then select one and now that is the current clipboard selection?

**Comment by Slackadays** [comment]
@niksingh710 you might have to make a script to do this because there's no built in way to do this with just dmenu.

however, that script might not be too hard to do because all you need is a way to parse JSON and then select the results.

here's a quick command I just whipped up that lets you select all current text clipboard contents and then send it to the main clipboard:

```sh 
cb st | jq -r '.[]' | dmenu | cb
```

this works by:
1. showing all current clipboard contents with `cb st` (the pipe makes it give JSON)
2. using jq to extract all values from the keys with `.[]` and then removing the quotes from them with `-r`
3. sending it to dmenu so you can select which one you want
4. sending the selection to cb to be copied

hope this helps! :)

**Reply by semi710** [reply]
![image](https://github.com/Slackadays/Clipboard/assets/60490474/36ad10da-a519-4d4f-8449-7f7a3b496bf7)

i get this error if there is a png in clipboard or some file.
currently i am using cliphist to acheive this would love if that is possible using this.

**Reply by Slackadays** [reply]
@niksingh710 fixed in latest commit https://github.com/Slackadays/Clipboard/commit/291300d42efc0033e6b54e58042f38808591da3d, now it won't show that data in JSON but rather what kind of content it is

also, it looks cooler now because of the fancy format that binary content gets

**Reply by semi710** [reply]
Thanks buddy ...... it seems last release was 5 days ago so when to expect this to be added in the release? currently using `clipboard-bin` from AUR

**Reply by Slackadays** [reply]
The next release might come in a week or two

**Comment by Slackadays** [comment]
@niksingh710 next release coming out next Saturday (probably)

also look out for a really cool feature (OSC 52) that will also be in 0.8.0 :)

**Reply by Slackadays** [reply]
@niksingh710 the new release is out! the aur should be updated very soon

**Reply by semi710** [reply]
received the new update mate but now using the suggested command above results in a json string.

how can i pass cb history content in a dmenu and select it as the current selection for cb?

**Reply by Slackadays** [reply]
> received the new update mate but now using the suggested command above results in a json string.
> 
> how can i pass cb history content in a dmenu and select it as the current selection for cb?

@niksingh710 are you sure you're using the right command? because when I do `cb st | jq -r '.[]' | dmenu | cb` with the latest version it works the same as before, what does it look like on your end?

**Reply by semi710** [reply]
Here is a verbose of my observations.

I am using Arch with wayland (wl-copy is available)

```bash
cb copy hi 			# This adds hi in the clipboard list
cb st				# Prints the `hi` entry in cb
cb copy "Hello"		# This should adds Hello in the clipboard list
cb copy "Hello2"	# This should adds Hello2 in the clipboard list
cb st				# This should show 3 entries but only the last entry is shown.
```

```bash
cb copy /path/to/img.png	# This copies the image and lists out the image file name in `cb st`
							# but in dmenu (rofi -dmenu -i) it shows the json array in every line.
cb st | jq -r '.[]' | rofi -dmenu -i
```
![image](https://github.com/Slackadays/Clipboard/assets/60490474/297211e1-6caa-4c76-ae18-293e1f290005)


`cb history` command lists all what i had copied till now but `cb st` only lists the last entry.

### Issues

`cb st` only listing last entry
`cb st` if passed in jq gives a full json object for the raw files.

Piping cb command after dmenu gives this
![image](https://github.com/Slackadays/Clipboard/assets/60490474/33f2d3ab-f9ac-4c7b-8891-3a6d1f01c316)

**Reply by semi710** [reply]
can ensure same behaviour for dmenu also. (Just tested)

**Reply by Slackadays** [reply]
ok, so it looks like you can fix `cb st if passed in jq gives a full json object for the raw files` by doing `cb st | jq -r '.[] | select(type=="string")'` to only select strings from the json. 

for `cb history`, you can do `cb hs | jq -r '.[].content | select(type=="string")'` to choose the last entries in the history (text only again).

however, I don't know if `cb st only listing last entry` can be fixed because I can't think of a good UX that would make this work, but if you have an idea, I would be happy to add it

**Reply by semi710** [reply]
Okay!! 
So this does kinda work but does not fulfill what I desire for.

Let's take an example application, Telegram.
### What I want to achieve?
```bash
cb copy /path/to/img.png 		# Copying the img in system clipboard
``` 
Then using `Ctrl+v` to paste that in the gui of telegram opened chat.
Now This is possible if I use wl-copy for the raw file but not if using cb.

And to get text pasted using `Ctrl+v` I do
```
 cb hs | jq -r '.[].content | select(type=="string")' | rofi -dmenu -i | tee > (wl-copy) | cb
```

For the UX part I observed that if using `cb st` in continues succession it sometimes messes up the terminal prompt and spaces.

In my opinion if UX is considered only on cli part then it can be a bit spacious as help menu feels much compact. This can have it's own discussion topic and i guess creating it a pinned topic or issue for community opinion will be a great option.

**Reply by Slackadays** [reply]
so it turns out what you want is impossible using 0.8.0, but is super easy if you can compile cb yourself. what you want to do is follow the compile instructions in the readme but BEFORE you build cb, make this change in `src/cb/src/actions/status.cpp`:

```
129 - printf("            \"name\": \"%s\",\n", entry.path().filename().string().data());
129 + printf("            \"name\": \"%s\",\n", entry.path().string().data());
```

then build and then you can use this command to select the files directly:

`cb st | jq -r '.[] | if type=="array" then .[].name else . end'`

for history, it's the same idea but make this change in `src/cb/src/actions/history.cpp` instead:

```
198 - printf("                \"name\": \"%s\",\n", entry.filename().string().data());
198 + printf("                \"name\": \"%s\",\n", entry.string().data());
```

then you can use this command to select from history:

`cb hs | jq -r '.[].content | if type=="array" then .[].name else . end'` 

I think rofi might offer an option so you don't see the long ugly filepaths every time you use dmenu, but it should work either way

I can try adding multiple entries to cb status for the next version as well as the full filepaths in the json, so stay tuned for that :)

**Reply by Slackadays** [reply]
@niksingh710 are you still using your dmenu setup for this?

**Comment by darukutsu** [comment]
This has also img preview, not sure if it handles any kind of data. I overcomplicated idea in previous edits. I just found about `cb ignore` . 
For example to ignore clips with less than 4chars `cb ignore '^.{0,4}$'`. This doesn't repeat content of clipboard rather switches it around + adds smol image icon near filename.
```
    clip_text_id=$(
    cb hs |
      jq -c '.[].content | if type=="array" then .[].path else if type=="object" then .path else . end end' |
      while read -r entry; do
        removed_quotes=$(echo $entry | sed 's/^"\(.*\)"$/\1/')
        if [ -f "$removed_quotes" ]; then
          echo -e "$removed_quotes\x00icon\x1f$removed_quotes"
        else
          echo "$removed_quotes"
        fi
      done |
      rofi -dmenu -format i -p "Clipboard:" -i
  )
  cb hs $clip_text_id
```
